### DISCUSSION TITLE
question: how is solid.js different from knockout.js?

### DISCUSSION THREAD

**Discussion by zpdDG4gta8XKpMCd** [solidjs/solid#562] [Q&A]
i am a old fan of knockout.js i see that a lot of ideas are correlating, can you please identify differences between solid and knockout other than the fact that unlike in knockout the templates and bindings to DOM are implemented in JSX in solid

**Comment by ryansolid** [comment]
I am big fan of Knockout. I used it in production on a large app for 8 years. I wrote a bit here: https://www.solidjs.com/docs/1.0.0#knockout.js

But mostly I sought to modernize Knockout taking some lessons on predictable execution/batching from MobX. Solid's reactivity is setup by default to read/write separation which is a bit more explicit. A lot of the design aspects were to avoid what I felt were some of the less controlled aspects of Knockout.

JSX compilation are a big part because they let us basically do a form of loop unwinding. Knockout would walk the DOM tree to attach bindings parsing strings and looking up bindings. In Solid we look at the structure of the code and transform the view into JavaScript. So there is less iteration and we leverage JavaScript scope instead of a virtual scope from the template.

But high level it is the same concepts. I've said many times that Knockout had it mostly figured out 10 years ago. We've address the typical performance shortcomings at creation and get to leverage all the powerful granular data modeling and update performance.
